feat(bh-add-item)

lomamech opened the pull request:
Implementation of bh-add-item component, using this component in all module who used Add Bouton
- Patient Invoice
- Stock Entry
- Stock Exit
- Purchase Order
And Complexe Voucher

closes #2095

lomamech committed:
Resolve conflict

lomamech committed:
Resolve conflict with stock adjustement

jniles commented on client/src/js/components/bhAddItem.js:
```diff
@@ -0,0 +1,35 @@
+    $ctrl.name = $ctrl.name || 'AddItemForm';
+  };
+
+  // fires the onSelectCallback bound to the component boundary
```
I don't think there is a `onSelectCallback`.  Could you remove this comment or update it?

jniles commented on client/src/modules/invoices/patientInvoice.html:
```diff
@@ -110,25 +110,13 @@
+         <!-- "Component Add number of grid rows" input-group -->
+          <bh-add-item
+            item-increment = "PatientInvoiceCtrl.itemIncrement"
+            name = "add_item_invoice"
```
Since this isn't ever going to be submitted as part of a form, there is no need for a `name` binding.  `name` is only useful for form validation or submission to servers.

jniles commented on client/src/js/components/bhAddItem.js:
```diff
@@ -0,0 +1,35 @@
+    transclude  : true,
+    bindings    : {
+      itemIncrement    : '<',
+      recipient        : '<',
```
I don't see the value of `recipient` ... it looks like just another way to `disable` the input.  Since you already have the `disable` parameter, I propose you remove `recipient`.

jniles commented on client/src/modules/purchases/create/create.html:
```diff
@@ -134,23 +134,13 @@
-            ng-disabled="!PurchaseCtrl.supplier">
-          </div>
+        <bh-add-item
+          item-increment = "PurchaseCtrl.itemIncrement"
```
In my view, the component should handle this itself - there is no reason for the parent controller to care about the item increment.  That will clean up the component bindings and the parent controllers significantly.

jniles commented on client/src/modules/templates/bhAddItem.tmpl.html:
```diff
@@ -0,0 +1,20 @@
+<div ng-form="{{ $ctrl.name }}" bh-add-item ng-model-options="{ updateOn: 'default' }">  
```
There's no need to make an `ngForm` here ... we don't perform any form validation.  `ng-model-options` can also be removed (since it only makes sense to define on an `ngForm` or `ngModel`.

jniles reviewed (changes requested):
@lomamech, this is about half way there.

The skeleton of this component is good, but we can go further.  For example, the `recipient` and `disable` bindings are essentially the same: I propose you keep `disable` since it describes what will happen to the component if the binding is `true`.

Ideally, the basic component should look something like this:
```html
<bh-add-item callback="SomeCtrl.onChange(numItems)"></bh-add-item>
```

That is all we need for most modules.  The basic implementation can be optionally extended by specifying an initial `item-increment` and/or `disable` parameter.  The item increment can be a simple `@` binding that you parse with a call to `Number()`.  It will default to 1 if nothing is passed in.

This will make the component much simpler.  At the moment, it looks quite complicated!

lomamech committed:
perf(bhAddItem)
Perf component bh-add-item
- Use only option disable instead recipient and remove the last one
- Fix Comment and use the function callback instead instead function
  onChange
- Remove not necessary function name, item-increment and set this in
  component side
- Remove not necessary ng-form in component Templates

lomamech committed:
perf(bh-add-item)
- clean the code, remove some property not necessary for the component
- remove the $inject and use & instead &? for callback Property

lomamech committed:
perf(bhAddItems): using good synthaxe for component property

jniles commented on client/src/js/components/bhAddItem.js:
```diff
@@ -0,0 +1,35 @@
+  .component('bhAddItem', {
+    templateUrl : 'modules/templates/bhAddItem.tmpl.html',
+    controller  : addItemController,
+    transclude  : true,
```
Transclusion is not necessary for this component.

jniles commented on client/src/js/components/bhAddItem.js:
```diff
@@ -0,0 +1,35 @@
+    },
+  });
+
+addItemController.$inject = [];
```
You can go ahead and remove this line if we never `$inject` everything.

jniles commented on client/src/js/components/bhAddItem.js:
```diff
@@ -0,0 +1,35 @@
+    transclude  : true,
+    bindings    : {
+      disable          : '<',
+      callback         : '&?',
```
This should be required (`&` instead of `&?`).  The module will break without it!

jniles reviewed (approved):
@lomamech, just a few clean up issues.  Otherwise, this looks perfect :+1:

jniles commented on client/src/js/components/bhAddItem.js:
```diff
@@ -0,0 +1,32 @@
+    controller  : addItemController,
+    bindings    : {
+      disable     : '<',
+      Callback    : '&',
```
I think this should not be capital - that may be what is breaking your tests.
